### PR TITLE
Add `amp-story-auto-ads` front end

### PR DESCRIPTION
**Pull request by calebcordry**
This is a lot of files but most are static. I added a few more JSON payloads, and the actual example. I wasn't sure the best way to document as a lot of the struggles for publishers so far have been with the server side part.

A few things: 

I was having trouble getting the front && backend to run at the same time to test. Is there a good way to do this?

This is throwing CORS errors (I assume b/c localhost) Is there a way to avoid this?

This is a bit different in that it needs server side code to work. Do you think that we should put an example of this somewhere, and if so where would you recommend?

edit: (accidentally hit submit early :) ) 
I would like the demo of the story to run on the page if possible, but I don't need a page by page example. Is this possible?

### CODE REVIEW

**Comment by sebastianbenz** [comment]
`$ gulp backend:watch` should run the go backend locally (and watch for changes). There shouldn't be any CORS errors, can you share a log?

> I would like the demo of the story to run on the page if possible, but I don't need a page by page example. Is this possible?

This is not possible (yet), but it's easy to add via a flag. I can look into this. How would you like to embed the story into the page? Manually using amp-iframe?

**Comment by sebastianbenz** [comment]
Actually, it's already possible to disable the preview adding:

```
<!---{
  "hidePreview": true
}--->
```

in the metadata section.

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+          {
+              "ad-attributes": {
+                "type": "custom",
+                "data-url": "https://www.ampbyexample.com/json/amp-story-auto-ads"
+              }
+          }
+        </script>
+
+        <!-- After the script tag then inline any templates that you wish to include.
+        These templates will be populated from by a JSON response from the
```
Comment (sebastianbenz) [inline src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: wording: "populated from by a JSON response from the"

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        server indicated by the `data-url` key above.
+        -->
+        <div>
```
Comment (sebastianbenz) [inline src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: Suggestion: split the code samples into two paragraphs, one for image one for video.

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html [hunk 3]
```diff
@@ -0,0 +1,335 @@
+<!--- {
+  "preview": "default"
+}--->
+
+<!--
+  ## Introduction
+  In an AMP story, you cannot put an amp-ad directly onto the page, instead, all
```
Comment (sebastianbenz) [inline src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: nit: please surround all AMP extension names with `.

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+              width="400"
+              height="750"
+              poster="{{poster}}"
+              layout="fill">
+              <source src="{{videoSrc}}" type="video/mp4">
+            </amp-video>
+          </template>
+        </div>
+      </amp-story-auto-ads>
+
```
Comment (sebastianbenz) [inline src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: I assume you'd want to embed the story here. You should be able to declare the embed via amp-iframe inside the documentation block:

```
This is how the story looks....

<!---
<div class="abe-story-preview">
<amp-iframe media="(min-width: 650px)" src="/stories/introduction/amp_story_hello_world/preview/embed/"  width="440" height="732" sandbox="allow-scripts allow-popups allow-same-origin" ><div placeholder></div></amp-iframe>
</div>
-->
```

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html (hunk 1 above)
Comment (calebcordry) [reply src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: reworded

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html (hunk 2 above)
Comment (calebcordry) [reply src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: done.

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html (hunk 3 above)
Comment (calebcordry) [reply src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: done.

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html (hunk 4 above)
Comment (calebcordry) [reply src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: I am having trouble getting this to work. Can you PTAL and tell me what I am doing wrong. (right now it is pointing at the existing story and I still can't get it to render)

**Comment by calebcordry** [comment]
re: CORS
https://imgur.com/kKhyKbS

File: src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html (hunk 4 above)
Comment (sebastianbenz) [reply src/amp-ads/50_AMP_Story_Ads/amp-story-auto-ads.html]: Sorry about that, you're hitting a few (non-documented) edge cases. This is what you have to do:

```
      <!-- See the full example below:
      <div class="abe-story-preview">
          <amp-iframe media="(min-width: 650px)"
              src="https://amp-by-example-api.appspot.com/amp-ads/amp_story_ads/amp-story-auto-ads/index.html"
              width="440" height="732" sandbox="allow-scripts allow-popups allow-same-origin">
            <div placeholder></div>
          </amp-iframe>
        </div> -->
      <!-- -->
```

The last `  <!-- -->` is needed to not show the following tag as a code section.

**Comment by calebcordry** [comment]
The ads aren't there due to the CORS errors, I dug in a bit and it looks like I am getting a 301 for the initial request to `/json/amp-story-auto-ads` at which point it loses the `__amp_source_origin` part of the query string. I think because of that it is not returning the right headers. Am I doing something goofy with the backend endpoint?

image

**Comment by calebcordry** [comment]
Talked offline. Moved to `stories/monetization`. CORS errors should be fixed.

**Comment by sebastianbenz** [comment]
Awesome! Almost there - can you please fix the validation errors:

screen shot 2018-04-17 at 09 37 08

You can ignore the error about amp-story-auto-ad, that's been taken care of in the PR I've cc'd you.

**Comment by calebcordry** [comment]
done.

**Comment by sebastianbenz** [comment]
@calebcordry please let me know when this is ready to commit.

**Comment by calebcordry** [comment]
Will do. There was a bug with `amp-mustache` scrubbing the `src` attribute on `<source />`. The fix is in, but waiting for it to hit prod.

**Comment by calebcordry** [comment]
cc/ @jasti for visibility

**Comment by sebastianbenz** [comment]
Any updates on this?

**Comment by calebcordry** [comment]
The fix is here: ampproject/amphtml#14762  Still just waiting for it to move to prod.

**Comment by newmuis** [comment]
@calebcordry can we merge in the changes from ampproject/amphtml#15093 then merge this PR?  I'd like to use this ABE example as a test case in validator tests

**Comment by calebcordry** [comment]
Yes! The bugfix is finally in prod. Will copy over your changes today.

**Comment by calebcordry** [comment]
@sebastianbenz I think this is ready to merge, let me know if anything else is needed.

**Pull request merged**